toolkit/scripts/help.mk

```diff
@@ -0,0 +1,106 @@
+# 
+#     ##help:TAG:CONTENT
+#
+# The format of the content differs by tag. Here are the main tags used:
+#
+#     ---------------------------------------------------------------------
+#     target
+#     ---------------------------------------------------------------------
+#     Describes a Makefile target intended for direct use by consumers of
+#     the toolkit. For consistency's sake, it recommended that the description
+#     of the target be terminated by a period (.).
+#
+#     Format:
+#         ##help:target:TARGET-NAME=DESCRIPTION-OF-TARGET
+#
+#     Example:
+#         ##help:target:MY_TARGET=My target.
+#
+#     ---------------------------------------------------------------------
+#     var
+#     ---------------------------------------------------------------------
+#     Describes a well-known variable that influences the Makefile and its
+#     targets in a meaningful (and supported) way.
+#
+#     Format:
+#         ##help:var:VAR-NAME:POSSIBLE-VALUES=DESCRIPTION-OF-VAR
+#
+#     For each variable, we document the possible values that it can take
+#     on (if enumerable), or otherwise use a human-readable 'metavar' that
+#     describes what the value should be.
+#
+#     Examples:
+#         ##help:var:MY_ENUM:{a,b,c}=My enum var
+#         ##help:var:MY_STRING:<path_to_relevant_thing>=Path to relevant thing
+#
+#     ---------------------------------------------------------------------
+#     example
+#     ---------------------------------------------------------------------
+#     Describes general example usage for the Makefile.
+#
+#     Format:
+#         ##help:example:EXAMPLE-TEXT-LINE-HERE
+#
+# There are also special metadata for the toplevel description included at
+# the start of the help output, as well as items for the headings used before
+# each section:
+#
+#     desc            - toplevel description
+#     target-heading  - heading displayed before the list of targets
+#     var-heading     - heading displayed before the list of variables
+#     example-heading - heading displayed before the list of examples
+# 
+# The 'help' target in the toplevel Makefile is responsible for extracting these
+# metadata comments, sorting them (where required), and rendering them into
+# sensible-looking output.
+#
+# The remainder of this file includes some of the special tags described above.
+#
+
+#
+# Top-level description.
+#
+
+##help:desc:[1mThis Makefile is the primary entry point for the Mariner Toolkit.[0m
+##help:desc:
+##help:desc:For full details, please consult the documentation under toolkit/docs.
+
+#
+# Headings.
+#
+
+##help:target-heading:
+##help:target-heading:[1;4;34mTargets:[0m
+
+##help:var-heading:
+##help:var-heading:[1;4;34mVariables:[0m
+
+##help:example-heading:
+##help:example-heading:[1;4;34mExamples:[0m
+
+#
+# Examples.
+#
+
+##help:example:Build only my-package.spec and its requirements:
+##help:example:    [3msudo make -j$(nproc) build-packages SRPM_PACK_LIST="my-package"[0m
```

Comment:
I think we should mention the `QUICK_REBUILD_*` set of arguments, as they were designed to reflect common build cases by changing the defaults of some of the more obscure flags under the hood. There's more details about their purpose inside `building.md` and `incremental_building.mk`, but if that's not enough, @dmcilvaney is an expert on them.

Reply:
I'll admit -- I'm not clear when I should and shouldn't use it. If i'm a human being and not an ADO pipeline, should I *always* use `QUICK_REBUILD=y`? If no, then some succinct pointers on when to use it would help. If yes, then is it really `I_AM_A_HUMAN_BEING_AND_WANT_MAKE_TO_GO_FASTER=y`? :) (Since we're using Makefile as a UI to the build, I'd strongly urge us to tune its defaults to the dev case rather than the pipeline case; the pipelines are going to explicitly specify exactly-the-right-set-of-args anyhow.)

Reply:
I've been tyring to find time to optimize the defaults for users, but features keep taking up my focus.

The big outstanding issue before I turn that on is fixing the toolchain build. Currently if you don't know to always set `CLEAN_TOOLCHAIN_CONTAINERS=n` every time you run `make clean` setting QUICK_REBUILD_TOOLCHAIN=y (which is set by the higher level QUICK_REBUILD=y flag) will hit you with a >1hr build penalty for almost every type of build as it regenerates the bootstrap container (which it might not even need...).

Reply:
(fyi, I have https://github.com/microsoft/CBL-Mariner/pull/5218 but it is now out-of-date. I've got a task to refresh this once we finish the parallel build upgrades).

Reply:
I completely understand the time constraints. It sounds like there's work to be done before certain semantics can be turned on by default... and I can appreciate that. But is there a way to separate that from the human interface? In other words, can we front-load agreement once and for all on what devs with a 100 or 200-level familiarity with the toolkit should put (or should *not* put) on the command line to request "fast local incremental builds"? Once that interface is set, we can presumably change what that maps to underneath the hood, in terms of which individual build features are enabled/disabled.

Reply:
Anyhow, I'm happy to chat about this broader discussion separately. For the purposes of this PR, @dmcilvaney , which examples -- if any -- of `QUICK_REBUILD*` would you recommend including in the help examples? I'm also completely okay pending that for a future PR if you'd prefer that we do so.

Reply:
I'm going to refresh the build docs, why don't I do this at the same time?

Reply:
That sounds great. I think you'll be able to compose better verbiage than I would for it anyhow.